renderdoc/driver/gl/wrappers/gl_texture_funcs.cpp

```diff
@@ -1372,22 +1372,60 @@ void WrappedOpenGL::glCopyImageSubData(GLuint srcName, GLenum srcTarget, GLint s
       {
         TextureData &dstData = m_Textures[dstrecord->GetResourceID()];
 
-        if(srcX == 0 || srcY == 0 || srcZ == 0 || dstX == 0 || dstY == 0 || dstZ == 0)
-        {
-          // we only support whole copies - sub-copies will not work correctly.
-          RDCASSERT(srcWidth == RDCMAX(1, srcData.width >> srcLevel));
-          RDCASSERT(srcHeight == RDCMAX(1, srcData.height >> srcLevel));
-          RDCASSERT(srcDepth == RDCMAX(1, srcData.depth >> srcLevel));
+        GLsizei srcLevelWidth = RDCMAX(1, srcData.width >> srcLevel);
+        GLsizei srcLevelHeight = RDCMAX(1, srcData.height >> srcLevel);
+        GLsizei srcLevelDepth = RDCMAX(1, srcData.depth >> srcLevel);
 
-          RDCASSERT(srcWidth == RDCMAX(1, dstData.width >> dstLevel));
-          RDCASSERT(srcHeight == RDCMAX(1, dstData.height >> dstLevel));
-          RDCASSERT(srcDepth == RDCMAX(1, dstData.depth >> dstLevel));
+        GLsizei dstLevelWidth = RDCMAX(1, dstData.width >> dstLevel);
+        GLsizei dstLevelHeight = RDCMAX(1, dstData.height >> dstLevel);
+        GLsizei dstLevelDepth = RDCMAX(1, dstData.depth >> dstLevel);
 
+        if(srcX == 0 && srcY == 0 && srcZ == 0 && dstX == 0 && dstY == 0 && dstZ == 0 &&
+           srcLevelWidth == dstLevelWidth && srcLevelHeight == dstLevelHeight &&
+           srcLevelDepth == dstLevelDepth)
+        {
           dstData.compressedData[dstLevel] = srcData.compressedData[srcLevel];
         }
         else
         {
-          RDCWARN("glCopyImageSubData doesn't support offsetted copies of compressed data");
+          size_t srcImageSize = GetCompressedByteSize(srcLevelWidth, srcLevelHeight, srcLevelDepth,
+                                                      srcData.internalFormat);
+          size_t dstImageSize = GetCompressedByteSize(dstLevelWidth, dstLevelHeight, dstLevelDepth,
+                                                      dstData.internalFormat);
+          std::tuple<size_t, size_t> srcBlockSize = GetCompressedBlockSize(srcData.internalFormat);
+          std::tuple<size_t, size_t> dstBlockSize = GetCompressedBlockSize(dstData.internalFormat);
+
+          RDCASSERT(srcWidth % std::get<0>(srcBlockSize) == 0);
+          RDCASSERT(srcWidth % std::get<0>(dstBlockSize) == 0);
+          RDCASSERT(srcHeight % std::get<1>(srcBlockSize) == 0);
+          RDCASSERT(srcHeight % std::get<1>(dstBlockSize) == 0);
+          for(size_t z = 0; z < (size_t)srcDepth; z++)
+          {
+            size_t srcLineSize = GetCompressedByteSize(
+                srcLevelWidth, (GLsizei)std::get<1>(srcBlockSize), 1, srcData.internalFormat);
+            size_t dstLineSize = GetCompressedByteSize(
+                dstLevelWidth, (GLsizei)std::get<1>(dstBlockSize), 1, dstData.internalFormat);
+            size_t srcOffset = srcImageSize * (srcZ + z) +
+                               srcLineSize * (srcY / (GLsizei)std::get<1>(srcBlockSize));
```

Comment:
This calculation seems wrong to me. `srcImageSize` is calculated as the size of all of the depth slices with `srcLevelDepth`, but it's being multiplied by the Z offset here which is iterating over the slices within that subregion. I think `srcImageSize` should be calculated with depth = 1 then so this offset calculation works? And similarly for `dstImageSize` in its use calculating `dstOffset`.

I think it would be better also to rename it to `srcSliceSize` or `srcLayerSize` to indicate that it's the size of only one z slice/layer.